### PR TITLE
Wait for an initial setWidth call before trying to render

### PR DESCRIPTION
**Pull request by cmdcolin**
This aims at fixing #1109 

It initializes self.width as undefined, and waits on a setWidth call before trying to render. It will throw if anything tries to access model.width before it is initialized, and users can check the LGV model.initialized getter

URL that can be used to check

http://localhost:3000/?config=test_data%2Fconfig_demo.json&session=eJztVVtv2jAU_ivIT5sEIQTKJW9tp7ZorEOAVm3ThExySLwmTmabW1H--46dLIRBq1WatJdJ8OBz-75z_DlnTziNgbjkHja1KUjJEl5zbMdu2H38zRzb7diuY1uD9uALqZOYioBx4tp14gu6AfHAfBUSt93v1MmawUYS9-ueMB9L2ljs8fPwR4h5PpNpRHf3OVgYtAZoVLtUn0aMAxW3wJMYPmEJ9CTLpQQ13hK3ZTu9rnVhd_rdbqdv951enSzSMQjjLAuDP4EAuefwApYFUguLSUWFMpSBIy-nM3Au7K6DyQLWICSgcUkjCXVCpYR4ER3xzL4hU0G9x0prLbt1M_RvnPDQxWXEAh4DV3Kmg9ERAgtCBEa0OvESvmTBSlCFLDGcB3Ek5sidKYUcMHzMIlilebJb4DR6H5x381UjHR-AqnEHkJZ9CrI_n2KaGfqnJOapiZtvd08YVghj-v5q0q6l1FuwpPbm_vbDaPK2VqbUEh7tankeOZ6fHlc-QRygRxUEidhpm6mojdSnKRapEL2i8WVhxGum8Sjxyl5WgukrUSqVbrMp2xaN6VPC6UZaXhI3vy9EspFgJSJoBkZLsqnRm_JxIdpNAdSXc92tPlpLOjfNN2zLsdrzmKYp-FbZloXYJKsTxn3YavDoHxLBPyNZlmm9Ih-h5YI08kudFKaT2y4dmDddBzdA1UrAmfAzzsygTe_H1wk-EBrAsSovetPGzKfT3kGUJ7EHZXZeEOaZtOfVKXnqFcGvl2gl-S_otML7oFe5WvxX9KsU_bukK2M9J9QzXlNDQgSe0itAWxkPMFj3KsNk8yuDuEqsILflDyS36K97yHy4wyY0WrEIDqaPmK8XW-kyAp0ayETMcmohQxDhhcyjEclBjJ5HdAGRrGJf44YAoVdeUVAT2DA_AGVGUK00qyKVr-_5iPI13j0X8mCAiH6QUURTs_r2OKkli5DVDLZKjy7f5EcbXI-Zeoqt4eGPuL7gy8zngOMB1TfkuJ25px_hPst-AlXi9Kg

Before

PileupRenderer.ts:195 {blockKey: "{hg19}1:8,801..9,600-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:9,601..10,400-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:10,401..11,200-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:11,201..12,000-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:8,001..9,600-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:9,601..11,200-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:11,201..12,800-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:12,801..14,400-0"}

After
PileupRenderer.ts:195 {blockKey: "{hg19}1:8,001..9,600-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:9,601..11,200-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:11,201..12,800-0"}
PileupRenderer.ts:195 {blockKey: "{hg19}1:12,801..14,400-0"}

### CODE REVIEW

File: packages/jbrowse-web/src/rootModel.ts [hunk 1]
```diff
@@ -276,6 +276,8 @@ export function createTestSession(snapshot = {}) {
     ...snapshot,
   })
   root.setHistory(UndoManager.create({}, { targetStore: root.session }))
+  // @ts-ignore
+  root.session.views.map(view => view.setWidth(800))
```
Comment (rbuels) [inline packages/jbrowse-web/src/rootModel.ts:280]: why are we manually setting the width here, can't we wait for the overall view to set it?

Comment (cmdcolin) [reply packages/jbrowse-web/src/rootModel.ts:280]: thjat particular line is for creating test snapshots, and the jest-dom does not do this for us

**Comment by cmdcolin** [comment]
>Why don't we just have width default to undefined instead of 800, and have the view not render if its width is undefined?

That is basically what this PR is doing? It has to manually make setWidth become called in jest-dom though hence some of the updates to the tests

**Comment by cmdcolin** [comment]
The update to packages/linear-genome-view/src/BasicTrack/blockBasedTrackModel.ts was also needed because the autorun starts running even despite the view not rendering it

**Comment by cmdcolin** [comment]
Note that this could also be addressed by making block width unrelated to the view width, and could address https://github.com/GMOD/jbrowse-components/issues/718 (or, that could be done separately, either way I personally find #718 is pretty bad when a lot of data is loaded)

**Comment by cmdcolin** [comment]
I made a hard-coded block width of 1000, because the fix of not rerendering when a drawer opens is a big benefit IMO

**Comment by cmdcolin** [comment]
I have still kept the need to manually set self.width though, because it does seem good to, when possible, not provide junk initial values

**Comment by cmdcolin** [comment]
Any thoughts on this?

**Comment by cmdcolin** [comment]
A similar fix was added to dotplot view for this also

**Comment by rbuels** [comment]
I like the hardcoded block-width idea. I don't like `volatileWidth` though. I think `width` just needs to default to undefined and code needs to deal with it being undefined if the view hasn't been set up by its container.

**Comment by cmdcolin** [comment]
Having it this way is an efficient way to root out any unexpected use of an undefined width, and force the user to consider how to deal with this situation rather than fumbling with undefined state (e.g. fail early, not kicking the can down the road). It helped in debugging issues in this PR for example.

I can revert if interested though

**Comment by rbuels** [comment]
That is a good point, failing fast is good.

Another concern I have (partially predating this PR) is that we have the `initialized` view, and then we have the width, which is also kind of needed by initialization. I kind of dislike both of these things, because as time goes on it gets less and less clear what it ought to mean to be `initialized`. Like, does the `self.volatileWidth !== undefined` need to be in the `initialized` view?

**Comment by cmdcolin** [comment]
Ya the fail fast and additionally, it just makes other "simple view types" that use self.width basically all have to consider that undefined case unless we cast it away which is kinda annoying

**Comment by cmdcolin** [comment]
>Like, does the self.volatileWidth !== undefined need to be in the initialized view?

Currently it is in the initialized view, because to me the view is not initialized until something has called setWidth. I maybe agree that it makes the initialized state a little ad-hoc (it could have a better defined name than just initialized perhaps), but I don't think it's necessarily wrong?

**Comment by rbuels** [comment]
No it's not wrong, I guess I'm just being cranky. Merging.

**Pull request merged**